// diagram/webpack.config.js
const path = require('path')
const pkg = require('./package.json')

function resolve (pathname) {
  return path.resolve(__dirname, pathname)
}

const config = {
  entry: {
    // 'demo/dist/index': resolve('./demo/src/index.js'),
    // 'demo/dist/mindmap': resolve('./demo/src/mindmap.js'),
    'demo/dist/tree': resolve('./demo/src/tree.js')
    // 'demo/dist/tree-dendrogram': resolve('./demo/src/tree-dendrogram.js'),
    // 'demo/dist/tree-indented': resolve('./demo/src/tree-indented.js'),
    // 'demo/dist/tree-layered-tidy': resolve('./demo/src/tree-layered-tidy.js'),
  },
  output: {
    filename: '[name].js',
    library: 'diagram',
    libraryTarget: 'umd',
    path: resolve('./')
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        loader: 'babel-loader',
        options: {
          presets: [
            // 'env',
            // 'es2015',
            'es2016'
            // 'es2017',
            // 'stage-0',
            // 'stage-1',
            // 'stage-2',
            // 'stage-3',
          ],
          plugins: [
            // 'transform-runtime'
          ]
        }
      }
    ]
  }
}

config.entry[`dist/${pkg.name}`] = resolve('./lib/index')

module.exports = config
